src/arch/cortex-m0/ctx_switch.S

```diff
@@ -27,29 +27,61 @@ EXC_RETURN_PSP:
   .word 0xFFFFFFFD
 
 .thumb_func
-/* r0 is top of user stack, r1 is heap base */
+/* r0 is top of user stack, r1 Process GOT */
 switch_to_user:
-  /* Load bottom of stack into Process Stack Pointer */
-  msr psp, r0
-
-  /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
-   * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
-  push {r4-r7}
-  mov r4,  r8
-  mov r5,  r9
-  mov r6, r10
-  mov r7, r11
-  push {r4-r7}
-
-  mov r9, r1
-  svc 0xff
-
-  /* These instructions are equivalent to the 32-bit "pop {r4-r11}" */
-  pop {r4-r7}
-  mov  r8, r4
-  mov  r9, r5
-  mov r10, r6
-  mov r11, r7
-  pop {r4-r7}
-
-  bx lr
+    /* Cortex-M0 can only push registers R0-R7 directly, so move R8-R11 to R0-R3.
+     * This is equivalent to the 32-bit "push {r4-r11}" instruction. */
+    push {r4-r7}
+    mov r4,  r8
+    mov r5,  r9
+    mov r6, r10
+    mov r7, r11
+    push {r4-r7}
+
+    /* Load non-hardware-stacked registers from Process stack */
+    ldmia r0!, {r4-r7}
+    mov r4,  r8
+    mov r5,  r9
+    mov r6, r10
+    mov r7, r11
+    ldmia r0!, {r4-r7}
+
+    /* Load bottom of stack into Process Stack Pointer */
+    msr psp, r0
+
+    /* Set PIC base pointer to the Process GOT */
+    mov r9, r1
```

Comment:
Well, one of those is definitely redundant.

One view is that, in principal, we only need to set r9 when the process starts, and the process should be able to do whatever it wants with r9. GCC compiled programs will never touch that register. But, you know... there is no particular reason to care from the kernel's perspective, and maybe some other compiler (or custom rolled assembly) will decide to reference the GOT some other way (maybe it knows it doesn't have a GOT, so it could actually use r9).

In fact, one way of implementing this strategy is just to pass in the base register as a third argument to _start (through the init callback), and let the process put it r9 (or wherever it wants)

On the other hand, setting r9 explicitly to the memory base on every context switch would, in theory, _might_ allow us to do some fancy relocation of running processes -- but I'm not sure it could even work given that processes can dynamically create pointers to their memory (i.e. at runtime) that don't need to go through the GOT, which we wouldn't be able to patch up.